[BACKEND-806] Modifications for internal build

xanec opened the pull request:
This modifies the build to:
1. fix the version of `joda-time` to `2.9.9` to ensure repeatable build
2. use (active) repositories configured in `settings.xml` (if available) for build

Commit:
modify dependency to fix version for joda-time.

Commit:
version variable for joda time.

Commit:
include build code to use m2 settings.xml repositories/credentials if
detected.

drcrallen commented on build.gradle:
```diff
@@ -12,7 +12,32 @@ subprojects {
 
     repositories {
-        jcenter()
+        File settingsFile = new File(m2SettingsFile.replace('~', System.env.HOME.toString()))
```
should only be at start of string, yes?

drcrallen commented on build.gradle:
```diff
@@ -12,7 +12,32 @@ subprojects {
-        jcenter()
+        File settingsFile = new File(m2SettingsFile.replace('~', System.env.HOME.toString()))
+        if (settingsFile.isFile()) {
+            println "Detected m2 settings file: ${settingsFile}"
```
does gradle not have a standard library that interfaces with maven?

xanec replied:
Apparently, the only way (that I know of) is to hard-code the settings (including credentials) into the build file to access custom repository.

drcrallen replied:
What does `mavenLocal()` not do  that `settings.xml` does?

drcrallen replied:
We really need to be using https://docs.gradle.org/current/userguide/dependency_management.html#sec:repositories if at all possible.

xanec replied:
I believe `mavenLocal()` only looks to the local maven cache (i.e.., `~/.m2/cache`) but does pull form external repositories

drcrallen commented on gradle.properties:
```diff
@@ -12,3 +12,6 @@ jacksonVersion=1.8.3
 mockitoVersion=1.8.5
+jodaTimeVersion=2.9.9
+
+m2SettingsFile=~/.m2/settings.xml
```
This is not really safe. many build environments don't pull from `~/.m2` and `~/.m2` is not guaranteed to be the maven directory.

xanec replied:
You can override the properties using the flag `-Pm2SettingsFile=<path-to-settings.xml>`, would this be okay?

drcrallen commented:
Suggest using standard libraries to interface with maven instead of writing our own.

Commit:
modified to use maven-settings plugin to read credentials from
settings.xml.

To specify credentials/custom repo, set
  -PrepoName=<id-of-server-in-settings.xml>
  -PrepoUrl=<URL of repo>

Note that there will be need for the builder to know in advance
the server ID and URL with this approach.

xanec commented:
Seems like there is definitely no support on Gradle-side to read configurations from `settings.xml` (refer to [this](https://issues.gradle.org/browse/GRADLE-1039) and [this](https://github.com/gradle/gradle/issues/1236)). This new commit gets pretty close by only requiring the builder to specify the `server id` in the `settings.xml` and the respective URL.

Commit:
change to use gradle.properties file to inject repo information.

dgolitsyn commented on build.gradle:
```diff
@@ -1,18 +1,31 @@
 
     repositories {
-        jcenter()
+        if (project.hasProperty("repoName") && project.hasProperty("repoUrl")) {
```
Where auth credentials for specified repo should be provided?

xanec replied:
The credentials will be read from the maven's `settings.xml` file. Specifically, the value of `repoName` should points to one of the `id` tag under `settings->servers->server`, then the plugin should use the corresponding crendentials
